### PR TITLE
Add BufferedSliceWriter from segmentation tools

### PR DESCRIPTION
**Pull request by MartinBuessemeyer**
This PR will add the BufferedSliceWriter from `segmentation-tools` to the `utils.py`. Therefore it will complete the first part of #68.

Is `utils.py` the right file or is it better to create a new file?

TODO: Testing

### CODE REVIEW

File: wkcuber/utils.py [hunk 1]
```diff
@@ -152,3 +152,84 @@ def time_stop(identifier):
...
+    def write_slice(self, z: int, data: np.ndarray):
+        """Takes in a slice in [y, x] shape, writes to WKW file."""
+
+        if len(self.buffer) == 0:
+            self.current_z = z
+            self.buffer_start_z = z
+
+        assert (
+            z == self.current_z
+        ), "({}) Slices have to be written sequentially!".format(getpid())
```
Comment (normanrz) [inline wkcuber/utils.py]: I don't think we need to print the pid in the logs

**Comment by MartinBuessemeyer** [comment]
How should I do the testing? Or can I skip them since this is just a copy of the class from the `segmentation tools`?

File: wkcuber/utils.py (hunk 1 above)
Comment (MartinBuessemeyer) [reply wkcuber/utils.py]: Done :grin:

**Comment by normanrz** [comment]
I think a test would be useful. Maybe something like:
- Write 13 slices -> nothing should be written to disk yet
- Write 19 more slices -> 32 slices on disk
- Write 3 more slices -> still only 32 slices on disk
- Close the writer -> 35 slices on disk

**Comment by MartinBuessemeyer** [comment]
Has someone an idea of why the test fails? Or to be precise: Why is there data on the disk?

**Comment by normanrz** [comment]
Have tried reproducing that issue locally or on one our servers?

**Comment by normanrz** [comment]
wkw will always return data, if you do `dataset.read(..)`. if there is no actual data, you'll get a bunch of zeros

**Comment by MartinBuessemeyer** [comment]
> wkw will always return data, if you do `dataset.read(..)`. if there is no actual data, you'll get a bunch of zeros

Thanks a lot.

Fixed the test accordingly.

Additionally, I made the `Magnification` of the `BufferedSliceWriter` to an optional argument because I think that we might want to write in another Magnification in the future.
Should be ready for the review now :)

**Comment by hotzenklotz** [comment]
@philippotto Bump. Review required.

File: tests/test_utils.py [hunk 2]
```diff
@@ -27,3 +30,49 @@ def test_get_regular_chunks_max_inclusive():
...
+                pass
+
+        for i in range(13, 32):
+            writer.write_slice(i, test_img)
+        with wkw.Dataset.open(dataset_path, wkw.Header(dtype)) as data:
+            read_data = data.read(origin, (24, 24, 32))
+            assert np.squeeze(read_data).shape == (24, 24, 32), "The read data should have the shape: (24, 24, 32) " \
+                                                                "but has a shape of: {}"\
+                                                                .format(np.squeeze(read_data).shape)
+            assert read_data.size == read_data[read_data.nonzero()].size, "The read data contains zeros while the " \
```
Comment (philippotto) [inline tests/test_utils.py]: Nice assertion!

File: tests/test_utils.py [hunk 3]
```diff
@@ -27,3 +30,49 @@ def test_get_regular_chunks_max_inclusive():
...
+            writer.write_slice(i, test_img)
+
+    with wkw.Dataset.open(dataset_path, wkw.Header(dtype)) as data:
+        read_data = data.read(origin, (24, 24, 35))
+        assert np.squeeze(read_data).shape == (24, 24, 35), "The read data should have the shape: (24, 24, 35) " \
+                                                                "but has a shape of: {}"\
+                                                                .format(np.squeeze(read_data).shape)
+        assert read_data.size == read_data[read_data.nonzero()].size, "The read data contains zeros while the " \
+                                                                          "written image has no zeros"
+
```
Comment (philippotto) [inline tests/test_utils.py]: Wouldn't it be useful to have an assertion which checks that `test_img` and `read_data` hold the same data?

File: wkcuber/utils.py [hunk 4]
```diff
@@ -152,3 +153,89 @@ def time_stop(identifier):
...
+        self.buffer_start_z = None
+
+    def write_slice(self, z: int, data: np.ndarray):
+        """Takes in a slice in [y, x] shape, writes to WKW file."""
+
+        if len(self.buffer) == 0:
+            self.current_z = z
+            self.buffer_start_z = z
+
+        assert z == self.current_z, "Slices have to be written sequentially!"
```
[the remark this answers is not in the document]
Comment (MartinBuessemeyer) [reply wkcuber/utils.py]: OK. I reverted the change :)

File: tests/test_utils.py (hunk 3 above)
Comment (MartinBuessemeyer) [reply tests/test_utils.py]: Just added it :D

**Comment by MartinBuessemeyer** [comment]
Should I merge?

**Comment by philippotto** [comment]
Sure!

**Pull request merged**